bob-apps/bin/brt.rs

```diff
@@ -250,12 +367,216 @@ fn try_main() -> AnyResult<()> {
+}
+
+fn validate_blob(path: &Path) -> AnyResult<()> {
+    let mut reader = BlobReader::from_path(&path)?;
+    reader.read_header()?;
+    while !reader.is_eof() {
+        reader.read_record()?;
+    }
+    Ok(())
+}
+
+fn validate_recursive<F>(path: &str, suffix: &str, mut validate: F) -> AnyResult<Vec<String>>
+where
+    F: FnMut(&Path) -> AnyResult<()> + Clone,
+{
+    log::info!("Start validation");
+    let mut result = vec![];
+    let mut count = 0;
+    let mut last_log = Instant::now();
+    for entry in walkdir::WalkDir::new(path)
+        .into_iter()
+        .filter_map(|res| match res {
+            Ok(entry) => Some(entry),
+            Err(err) => {
+                log::error!("[{}] walkdir error: {}", path, err);
+                None
+            }
+        })
+    {
+        let entry_path = entry.path().as_os_str().to_str().unwrap().to_string();
+        if entry.file_type().is_file() && entry_path.ends_with(&suffix) {
+            match validate(entry_path.as_ref()) {
+                Ok(_) => {}
+                Err(err) => {
+                    log::error!("[{}] validation error: {}", entry_path, err);
+                    result.push(entry_path.to_string());
+                }
+            }
+            count += 1;
+            if last_log.elapsed() >= Duration::from_secs(1) {
+                log::info!("{} files validated", count);
+                last_log = Instant::now();
+            }
+        }
+    }
+    Ok(result)
+}
+
+fn ask_confirmation(prompt: &str) -> AnyResult<bool> {
+    let mut stdout = std::io::stdout();
+    let stdin = std::io::stdin();
+    loop {
+        let mut buf = String::new();
+        write!(stdout, "{} (y/N): ", prompt)?;
+        stdout.flush()?;
+        stdin.read_line(&mut buf)?;
+        let cmd = buf.trim();
+        match cmd {
+            "y" | "Y" => return Ok(true),
+            "n" | "N" => return Ok(false),
+            _ => writeln!(stdout, "Unknown command, try again")?,
+        }
+    }
+}
+
+fn move_and_recover_blob<P, Q>(
+    invalid_path: P,
+    backup_path: Q,
+    validate_every: usize,
+) -> AnyResult<()>
+where
+    P: AsRef<Path>,
+    Q: AsRef<Path>,
+{
+    std::fs::rename(&invalid_path, &backup_path).with_context(|| "move invalid blob")?;
+
+    log::info!("backup saved to {:?}", backup_path.as_ref());
+    recovery_blob(&backup_path, &invalid_path, validate_every)?;
+    Ok(())
+}
+
+fn get_backup_path(path: &str, suffix: &str) -> AnyResult<PathBuf> {
+    let mut i = 0;
+    loop {
+        let path = if i > 0 {
+            PathBuf::from_str(&format!("{}{}.{}", path, suffix, i))
+        } else {
+            PathBuf::from_str(&path)
+        }?;
+        if !path.exists() {
+            return Ok(path);
+        }
+        i += 1;
+    }
+}
+
+fn validate_blob_command(settings: &ValidateBlobSettings) -> AnyResult<()> {
+    let result = validate_recursive(&settings.path, &settings.blob_suffix, validate_blob)?;
+
+    if !result.is_empty() {
```

Comment:
`validate_blob_command` and `validate_index_command` have plenty in common :) IMO you can extract part with printing blobs (`print_corrupted(v: Vec<..>, name: &str)`, so it's like `print_invalid(result, "blobs")`). Probably same for `print_fails` (or maybe they are the same function). An optionally there can be some function like `collect_fails`, but maybe there is too much to cusomtize between two functions

Reply:
Done